### PR TITLE
Fixed #1615 modify TestCase for WebGL

### PR DESCRIPTION
**Pull request by dingpinglv**
modify TestCase for WebGL

### CODE REVIEW

File: tests/EffectsTest/EffectsTest.js [hunk 1]
```diff
@@ -333,7 +333,7 @@ var TurnOffTilesTest = EffecstsBaseLayer.extend({
         return "a = cc.TurnOffTiles.create(duration, gridSize, seed)";
     },
     getEffect:function(duration) {
-        var action = cc.TurnOffTiles.create( duration, cc.size(16,12), 20);
+        var action = cc.TurnOffTiles.create( duration, cc.SizeMake(48,32), 25);
```
Comment (ricardoquesada) [inline tests/EffectsTest/EffectsTest.js]: it should be `cc.size()` and not `cc.SizeMake`. Otherwise it won't work on JSB

File: tests/MotionStreakTest/MotionStreakTest.js [hunk 2]
```diff
@@ -0,0 +1,252 @@
+/****************************************************************************
+ Copyright (c) 2010-2013 cocos2d-x.org
+ Copyright (c) 2008-2010 Ricardo Quesada
+ Copyright (c) 2011      Zynga Inc.
```
Comment (ricardoquesada) [inline tests/MotionStreakTest/MotionStreakTest.js]: Motion Streak demo was created by Jason Booth (slipster216).
Not by be or Zynga.

File: tests/OpenGLTest/OpenGLTest.js [hunk 3]
```diff
@@ -786,7 +782,8 @@ var ShaderRetroEffect = OpenGLTestLayer.extend({
             program.updateUniforms();
 
             var label = cc.LabelBMFont.create("RETRO EFFECT","res/fonts/west_england-64.fnt");
-            label.setShaderProgram( program );
```
Comment (ricardoquesada) [inline tests/OpenGLTest/OpenGLTest.js:789]: this line should be here. otherwise it will break JSB

File: tests/OpenGLTest/OpenGLTest.js [hunk 4]
```diff
@@ -1055,7 +1044,12 @@ var GLTexParamterTest = OpenGLTestLayer.extend({
         return "tests texParameter()";
     },
     getTexValues:function() {
-        gl.bindTexture(gl.TEXTURE_2D, null);
+        if(sys.platform === "browser"){
+            var texture2d = cc.TextureCache.getInstance().textureForKey(s_pathGrossini);
+            gl.bindTexture(gl.TEXTURE_2D, texture2d.getName());
+        } else {
+            gl.bindTexture(gl.TEXTURE_2D, null);
+        }
```
Comment (ricardoquesada) [inline tests/OpenGLTest/OpenGLTest.js:1117]: Why this comparison ? 
I think WebGL supports binding a null texture... it is part of the specification.

File: tests/ParticleTest/ParticleTest.js [hunk 5]
```diff
@@ -551,18 +553,19 @@ var DemoRotFlower = ParticleDemo.extend({
     onEnter:function () {
         this._super();
 
-        this._emitter = cc.ParticleSystem.createWithTotalParticles(150);
+        this._emitter = new cc.ParticleSystemQuad();
+        this._emitter.initWithTotalParticles((cc.renderContextType === cc.WEBGL) ? 300 : 150);
```
Comment (ricardoquesada) [inline tests/ParticleTest/ParticleTest.js]: this is not correct. You should never check for the context.
Instead you should check for: `if ("openg" in sys.capabilities)`
And also, use the `create` API, and not the `init` API which is confusing for the users.

File: tests/ParticleTest/ParticleTest.js [hunk 6]
```diff
@@ -760,12 +763,12 @@ var DemoModernArt = ParticleDemo.extend({
     onEnter:function () {
         this._super();
 
-        this._emitter = cc.ParticleSystem.createWithTotalParticles(200);
+        this._emitter = new cc.ParticleSystemQuad();
+        this._emitter.initWithTotalParticles((cc.renderContextType === cc.WEBGL) ? 1000 : 200);
```
Comment (ricardoquesada) [inline tests/ParticleTest/ParticleTest.js]: please see my previous comment.

File: tests/RenderTextureTest/RenderTextureTest.js [hunk 7]
```diff
@@ -24,12 +24,13 @@
...
 
 
-var sceneRenderTextureIdx = -1;
+var sceneIdx = -1;
 
 var RenderTextureBaseLayer = BaseTestLayer.extend({
-
-    ctor:function() {
-        this._super(cc.c4b(0,0,0,255), cc.c4b(98,99,117,255) );
+    ctor:function () {
```
Comment (ricardoquesada) [inline tests/RenderTextureTest/RenderTextureTest.js:29]: why?
this breaks JSB.

File: tests/ShaderTest/ShaderString.js [hunk 8]
```diff
@@ -0,0 +1,388 @@
+/****************************************************************************
```
Comment (ricardoquesada) [inline tests/ShaderTest/ShaderString.js]: I added many of the shadertests in the opengltest.
there is no need to create another test. Let's use OpenGL test for this.

File: tests/SpriteTest/SpriteTest.js [hunk 9]
```diff
@@ -52,17 +52,45 @@ var spriteFrameCache = cc.SpriteFrameCache.getInstance();
 // SpriteTestDemo
 //
 //------------------------------------------------------------------
-var SpriteTestDemo = BaseTestLayer.extend({
+var SpriteTestDemo = cc.LayerGradient.extend({
```
Comment (ricardoquesada) [inline tests/SpriteTest/SpriteTest.js]: It should be a subclass of BaseTestLayer, otherwise it can't be automated

File: tests/SpriteTest/SpriteTest.js [hunk 10]
```diff
@@ -79,17 +107,7 @@ var SpriteTestDemo = BaseTestLayer.extend({
         var s = new SpriteTestScene();
         s.addChild(previousSpriteTest());
         director.replaceScene(s);
-    },
-
-    // automation
```
Comment (ricardoquesada) [inline tests/SpriteTest/SpriteTest.js:84]: don't remove these lines... required for automation

File: tests/SpriteTest/SpriteTest.js [hunk 11]
```diff
@@ -711,93 +729,73 @@ var SpriteZVertex = SpriteTestDemo.extend({
...
-
-            // Avoid Z-fighting with menu and title
-            var menu = this.getChildByTag( BASE_TEST_MENU_TAG );
-            menu.setVertexZ(1);
-            var title = this.getChildByTag( BASE_TEST_TITLE_TAG );
-            title.setVertexZ(1);
-            var subtitle = this.getChildByTag( BASE_TEST_SUBTITLE_TAG );
-            subtitle.setVertexZ(1);
-
-        }
```
Comment (ricardoquesada) [inline tests/SpriteTest/SpriteTest.js:794]: why did you remove these lines? I added recently to test Z buffer

File: tests/SpriteTest/SpriteTest.js [hunk 12]
```diff
@@ -1136,10 +1110,10 @@ var SpriteAliased = SpriteTestDemo.extend({
         // This change will affect every sprite that uses the same texture
         // So sprite1 and sprite2 will be affected by this change
         //
-        if( sys.platform == 'browser' )  {
+        if (sys.platform == 'browser' && cc.renderContextType === cc.CANVAS) {
```
Comment (ricardoquesada) [inline tests/SpriteTest/SpriteTest.js]: don't use cc.renderContextType... we have sys.capabilities and 'opengl'... otherwise it will break JSB

File: tests/SpriteTest/SpriteTest.js [hunk 13]
```diff
@@ -1198,10 +1171,10 @@ var SpriteBatchNodeAliased = SpriteTestDemo.extend({
...
```
Comment (ricardoquesada) [inline tests/SpriteTest/SpriteTest.js]: same as my previous comment

File: tests/SpriteTest/SpriteTest.js [hunk 14]
```diff
@@ -1209,8 +1182,7 @@ var SpriteBatchNodeAliased = SpriteTestDemo.extend({
 
     },
     onExit:function () {
-        if( sys.platform == 'browser' )  {
-        } else {
+        if (sys.platform !== 'browser' || cc.renderContextType !== cc.CANVAS) {
```
Comment (ricardoquesada) [inline tests/SpriteTest/SpriteTest.js]: same as my previous comment

File: tests/SpriteTest/SpriteTest.js [hunk 15]
```diff
@@ -3421,6 +3374,8 @@ var SpriteBatchNodeReorderOneChild = SpriteTestDemo.extend({
 
         this._batchNode = aParent;
         //[[aParent texture] generateMipmap];
+        if(cc.renderContextType === cc.WEBGL)
+            aParent.getTexture().generateMipmap();
         this.addChild(aParent);
```
Comment (ricardoquesada) [inline tests/SpriteTest/SpriteTest.js:3391]: use `if('opengl' in sys.capabilities)` instead.

File: tests/TileMapTest/TileMapTest.js [hunk 16]
```diff
@@ -92,7 +92,8 @@ var TileMapTest = TileDemo.extend({
     ctor:function () {
         this._super();
         var map = cc.TileMapAtlas.create(s_tilesPng, s_levelMapTga, 16, 16);
-        map.getTexture().setAntiAliasTexParameters();
+        if (cc.renderContextType === cc.WEBGL)
```
Comment (ricardoquesada) [inline tests/TileMapTest/TileMapTest.js]: use `if('opengl' in sys.capabilities)` instead

File: tests/TileMapTest/TileMapTest.js [hunk 17]
```diff
@@ -293,7 +293,8 @@ var TMXReadWriteTest = TileDemo.extend({
         var s = map.getContentSize();
 
         var layer = map.getLayer("Layer 0");
-        //layer.getTexture().setAntiAliasTexParameters();
+        if (cc.renderContextType === cc.WEBGL)
```
Comment (ricardoquesada) [inline tests/TileMapTest/TileMapTest.js]: `if('opengl' in sys.capabilities)`

File: tests/TileMapTest/TileMapTest.js [hunk 18]
```diff
@@ -518,15 +519,17 @@ var TMXTilesetTest = TileDemo.extend({
         this.addChild(map, 0, TAG_TILE_MAP);
         var s = map.getContentSize();
 
-        /*var layer;
-         layer = map.getLayer("Layer 0");
-         layer.getTexture().setAntiAliasTexParameters();
+        if (cc.renderContextType === cc.WEBGL) {
```
Comment (ricardoquesada) [inline tests/TileMapTest/TileMapTest.js]: `if('opengl' in sys.capabilities)`

File: tests/TileMapTest/TileMapTest.js [hunk 19]
```diff
@@ -1124,7 +1144,8 @@ var TMXBug987 = TileDemo.extend({
         for (var i = 0, len = childs.length; i < len; i++) {
             node = childs[i];
             if (!node) break;
-            //node.getTexture().setAntiAliasTexParameters();
+            if (cc.renderContextType === cc.WEBGL)
```
Comment (ricardoquesada) [inline tests/TileMapTest/TileMapTest.js]: `if('opengl' in sys.capabilities)`

File: tests/main.js [hunk 20]
```diff
@@ -56,3 +56,90 @@ var cocos2dApp = cc.Application.extend({
...
+
+var s_rcVisible = cc.RectZero();
+var s_ptCenter = cc.PointZero();
+var s_ptTop = cc.PointZero();
+var s_ptTopRight = cc.PointZero();
+var s_ptRight = cc.PointZero();
+var s_ptBottomRight = cc.PointZero();
+var s_ptBottom = cc.PointZero();
+var s_ptLeft = cc.PointZero();
+var s_ptTopLeft = cc.PointZero();
```
Comment (ricardoquesada) [inline tests/main.js]: I think `cc.RectZero()` and `cc.PointZero()` don't exist in JSB.
Instead use:
`cc.POINT_ZERO` and `cc.rect(0,0,0,0)`

File: tests/res/Shaders/example_ColorBars.fsh [hunk 21]
```diff
@@ -21,9 +21,17 @@ void main(void)
 	colors[7] = vec4(1,0.5,0,1);
 	colors[8] = vec4(1,0.5,0.5,1);
 	colors[9] = vec4(0.5,0.5,1,1);
+	vec4 selColor = vec4(1,0,0,1);
 
 	// inline to prevent "float" loss and keep using lowp
 //	int y = int( mod(( (gl_FragCoord.y+gl_FragCoord.x)*mod(CC_Time[0],5.0)) / 10.0, 10.0 ) );
-	int y = int( mod( CC_Time[3] + (gl_FragCoord.y + gl_FragCoord.x) / 10.0, 10.0 ) );
```
Comment (ricardoquesada) [inline tests/res/Shaders/example_ColorBars.fsh:27]: why did you remove those lines?

File: games/WatermelonWithMe/watermelon_with_me.js [hunk 22]
```diff
@@ -192,7 +192,7 @@ var GameLayer = cc.LayerGradient.extend({
 
         // coin only needed to obtain the texture for the Batch Node
         var coin = cc.Sprite.createWithSpriteFrameName("coin01.png");
-        this._batch = cc.SpriteBatchNode.createWithTexture(coin.getTexture(), 100);
+        this._batch = cc.Node.create(); //cc.SpriteBatchNode.createWithTexture(coin.getTexture(), 100);
```
Comment (ricardoquesada) [inline games/WatermelonWithMe/watermelon_with_me.js]: why? `SpriteBatchNode` objects are much faster. This might reduce the speed of the game.

File: tests/ActionsTest/ActionsTest.js [hunk 23]
```diff
@@ -1675,20 +1675,32 @@ var ActionCardinalSpline = ActionsDemo.extend({
...
-        cc.drawingUtil.drawCardinalSpline(this._array, 0, 100);
-        context.restore();
-
-        var s = director.getWinSize();
-
-        context.save();
-        context.translate(s.width / 2 - apPoint.x ,  apPoint.y - 50);
-        cc.drawingUtil.drawCardinalSpline(this._array, 1, 100);
-        context.restore();
+        if(cc.renderContextType === cc.CANVAS){
```
Comment (ricardoquesada) [inline tests/ActionsTest/ActionsTest.js]: let's use `if('opengl' in sys.capabilities)`

File: tests/ActionsTest/ActionsTest.js [hunk 24]
```diff
@@ -1830,17 +1842,23 @@ var ActionCatmullRom = ActionsDemo.extend({
...
-        context.save();
-        context.translate(50 - apPoint.x, apPoint.y - 50);
-        cc.drawingUtil.drawCatmullRom(this._array1, 50);
-        context.restore();
-
-        context.save();
-        context.translate(- apPoint.x, apPoint.y);
-        cc.drawingUtil.drawCatmullRom(this._array2, 50);
-        context.restore();
+        if(cc.renderContextType === cc.CANVAS){
```
Comment (ricardoquesada) [inline tests/ActionsTest/ActionsTest.js]: let's use `if('opengl' in sys.capabilities)`

File: tests/BaseTestLayer/BaseTestLayer.js [hunk 25]
```diff
@@ -236,7 +236,8 @@ var BaseTestLayer = cc.LayerGradient.extend({
             return array;
         } else {
             // implement a canvas-html5 readpixels
-            throw "readPixels Not implemented on canvas yet";
+            //throw "readPixels Not implemented on canvas yet";
+            return ctx.getImageData(x, y, w, h).data;
```
Comment (ricardoquesada) [inline tests/BaseTestLayer/BaseTestLayer.js:240]: great!

File: tests/ClippingNodeTest/ClippingNodeTest.js [hunk 26]
```diff
@@ -0,0 +1,732 @@
+/****************************************************************************
+ Copyright (c) 2010-2013 cocos2d-x.org
+ Copyright (c) 2008-2010 Ricardo Quesada
+ Copyright (c) 2011      Zynga Inc.
```
Comment (ricardoquesada) [inline tests/ClippingNodeTest/ClippingNodeTest.js]: This test was created by Pierre-David Bélanger.
Zynga and I were not involved in this code.

File: tests/CocosNodeTest/CocosNodeTest.js [hunk 27]
```diff
@@ -29,12 +29,15 @@ var TAG_SPRITE2 = 2;
 var TAG_SPRITE3 = 3;
 var TAG_SLIDER = 4;
 
-var nodeTestSceneIdx = -1;
+var sceneIdx = -1;
 var MAX_LAYER = 9;
 
-var TestNodeDemo = BaseTestLayer.extend({
+var TestNodeDemo = cc.Layer.extend({
```
Comment (ricardoquesada) [inline tests/CocosNodeTest/CocosNodeTest.js]: no. It should be a subclass of `BaseTestLayer`, otherwise it will break automation.

File: tests/CocosNodeTest/CocosNodeTest.js [hunk 28]
```diff
@@ -29,12 +29,15 @@ var TAG_SPRITE2 = 2;
...
     ctor:function () {
         this._super();
+
+        cc.associateWithNative(this, cc.Layer);
```
Comment (ricardoquesada) [inline tests/CocosNodeTest/CocosNodeTest.js]: and this is not needed, since `BaseTestLayer` does that.

File: tests/CocosNodeTest/CocosNodeTest.js [hunk 29]
```diff
@@ -42,31 +45,49 @@ var TestNodeDemo = BaseTestLayer.extend({
     subtitle:function () {
         return "";
     },
+    onEnter:function () {
```
Comment (ricardoquesada) [inline tests/CocosNodeTest/CocosNodeTest.js]: the menus are created by `BaseTestLayer`

File: tests/CocosNodeTest/CocosNodeTest.js [hunk 30]
```diff
@@ -42,31 +45,49 @@ var TestNodeDemo = BaseTestLayer.extend({
...
         s.addChild(nextNodeTest());
         director.replaceScene(s);
     },
-    onBackCallback:function (sender) {
+    backCallback:function (sender) {
         var s = new NodeTestScene();
         s.addChild(previousNodeTest());
         director.replaceScene(s);
-    },
-    // automation
```
Comment (ricardoquesada) [inline tests/CocosNodeTest/CocosNodeTest.js:61]: needed for automation

File: tests/CocosNodeTest/CocosNodeTest.js [hunk 31]
```diff
@@ -412,43 +433,43 @@ var CameraZoomTest = TestNodeDemo.extend({
...
         var sprite = cc.Sprite.create(s_pathGrossini);
         this.addChild(sprite, 0);
-        sprite.setPosition(cc.p(winSize.width / 4 * 1, winSize.height / 2));
-        var cam = sprite.getCamera();
-        if(cam.setEye)
-            cam.setEye(0, 0, 415);
-        if(cam.setCenter)
-            cam.setCenter(0,0,0);
+        sprite.setPosition(cc.p(winSize.width / 4, winSize.height / 2));
+        if (cc.renderContextType === cc.WEBGL) {
```
Comment (ricardoquesada) [inline tests/CocosNodeTest/CocosNodeTest.js]: it shall be: `if('opengl' in sys.capabilities)`

File: tests/CocosNodeTest/CocosNodeTest.js [hunk 32]
```diff
@@ -596,17 +617,17 @@ var ConvertToNode = TestNodeDemo.extend({
 // BoundingBox Test
 //
 var BoundingBoxTest = TestNodeDemo.extend({
-    ctor:function () {
+    init:function () {
```
Comment (ricardoquesada) [inline tests/CocosNodeTest/CocosNodeTest.js]: it should be `ctor` and not `init`

File: tests/CocosNodeTest/CocosNodeTest.js [hunk 33]
```diff
@@ -596,17 +617,17 @@ var ConvertToNode = TestNodeDemo.extend({
...
         this._super();
         var sprite = cc.Sprite.create(s_pathGrossini);
-        this.addChild( sprite );
-        sprite.setPosition(winSize.width/2, winSize.height/2);
+        this.addChild(sprite);
+        sprite.setPosition(winSize.width / 2, winSize.height / 2);
         var bb = sprite.getBoundingBox();
-        this.log('BoundingBox:');
+        cc.log('BoundingBox:');
```
Comment (ricardoquesada) [inline tests/CocosNodeTest/CocosNodeTest.js]: it should be `this.log` and not `cc.log`

**Comment by ricardoquesada** [comment]
thanks for making the pull request. 
A few comments:
- try not to mix cosmetic fixes (like indentation, whitespaces, etc), with real fixes in the same pull request.
- use `if('opengl' in sys.capabilities)` to check if opengl is enabled
- check everything on JSB. You will notice that your changes break it
- Tests should be a subclass of `BaseTestLayer` ir order to be automated

File: tests/OpenGLTest/OpenGLTest.js (hunk 4 above)
Comment (dingpinglv) [inline tests/OpenGLTest/OpenGLTest.js:1117]: Yes, WebGL supports binding a null texture, but if an attempt is made to call texParameteri function with no WebGLTexture bound , an INVALID_OPERATION error is generated.

File: tests/ShaderTest/ShaderString.js (hunk 8 above)
Comment (dingpinglv) [inline tests/ShaderTest/ShaderString.js]: I defined some shaders as string variable at first, so I  create this file. Now, cc.GLProgram supports load shader from text file. I will do this.

File: tests/SpriteTest/SpriteTest.js (hunk 10 above)
Comment (dingpinglv) [inline tests/SpriteTest/SpriteTest.js:84]: I think this was caused by merge automatically.

**Comment by dingpinglv** [comment]
Thank you very much,  riq!
I will correct my mistakes soon.

File: tests/res/Shaders/example_ColorBars.fsh (hunk 21 above)
Comment (dingpinglv) [inline tests/res/Shaders/example_ColorBars.fsh:27]: Because this code will fail at the shader-compilation stage, with the error message :'[]': index expression must be constant. 
I will find a better way to access array for webGL.
